File: backend/app/schemas/campaign.py
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class CampaignCreate(BaseModel):
    name: str = Field(..., min_length=3, max_length=255)
    description: Optional[str] = None

    test_source_type: Literal["scenario_library", "uploaded_dataset"] = "scenario_library"
    dataset_id: Optional[str] = None

    selected_models: list[str] = Field(default_factory=lambda: ["mock:mock-safe-model"])
    selected_scenario_ids: list[str] = Field(default_factory=list)
    selected_categories: list[str] = Field(default_factory=list)
    selected_mutations: list[str] = Field(default_factory=lambda: ["direct"])

    max_tests: int = Field(default=20, ge=1, le=200)


class CampaignResponse(BaseModel):
    id: int
    campaign_id: str

    name: str
    description: Optional[str]
    status: str

    test_source_type: str
    dataset_id: Optional[str]
    dataset_name: Optional[str]
    dataset_row_count: int

    selected_models: list[str]
    selected_scenario_ids: list[str]
    selected_categories: list[str]
    selected_mutations: list[str]

    max_tests: int
    total_tests: int
    completed_tests: int
    failed_tests: int
    critical_findings: int
    average_risk_score: int

    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]

    class Config:
        from_attributes = True


class CampaignListResponse(BaseModel):
    total: int
    limit: int
    offset: int
    items: list[CampaignResponse]


class CampaignRunResponse(BaseModel):
    campaign_id: str
    status: str
    message: str


class CampaignStatusResponse(BaseModel):
    campaign_id: str
    name: str
    status: str

    test_source_type: str
    dataset_id: Optional[str]
    dataset_name: Optional[str]
    dataset_row_count: int

    max_tests: int
    total_tests: int
    completed_tests: int
    failed_tests: int
    critical_findings: int
    average_risk_score: int

    progress_percent: float

    started_at: Optional[datetime]
    completed_at: Optional[datetime]


class CampaignResultResponse(BaseModel):
    id: int
    campaign_pk: int
    campaign_id: str

    scenario_pk: Optional[int]
    scenario_id: Optional[str]

    attack_name: Optional[str]
    attack_category: Optional[str]
    severity: Optional[str]
    owasp_category: Optional[str]

    model_name: str
    mutation_type: str

    input_prompt: str
    mutated_prompt: str

    sandbox_report: Optional[dict[str, Any]]
    model_response: Optional[dict[str, Any]]
    ai_evaluation: Optional[dict[str, Any]]
    risk_assessment: Optional[dict[str, Any]]

    risk_score: int
    final_status: Optional[str]
    error_message: Optional[str]

    created_at: datetime

    class Config:
        from_attributes = True


class CampaignResultsListResponse(BaseModel):
    campaign_id: str
    total: int
    items: list[CampaignResultResponse]

File: frontend/src/types/campaign.ts
export type CampaignCreateRequest = {
  name: string;
  description?: string;
  test_source_type: "scenario_library" | "uploaded_dataset";
  dataset_id?: string | null;
  selected_models: string[];
  selected_scenario_ids: string[];
  selected_categories: string[];
  selected_mutations: string[];
  max_tests: number;
};

export type CampaignResponse = {
  id: number;
  campaign_id: string;
  name: string;
  description?: string | null;
  status: string;

  test_source_type?: string;
  dataset_id?: string | null;
  dataset_name?: string | null;
  dataset_row_count?: number | null;

  selected_models: string[];
  selected_scenario_ids: string[];
  selected_categories: string[];
  selected_mutations: string[];

  max_tests: number;
  total_tests: number;
  completed_tests: number;
  failed_tests: number;
  critical_findings: number;
  average_risk_score: number;

  created_at: string;
  started_at?: string | null;
  completed_at?: string | null;
};

export type CampaignRunResponse = {
  campaign_id: string;
  status: string;
  message: string;
};

export type CampaignStatusResponse = {
  campaign_id: string;
  name: string;
  status: string;

  test_source_type?: string;
  dataset_id?: string | null;
  dataset_name?: string | null;
  dataset_row_count?: number | null;

  max_tests: number;
  total_tests: number;
  completed_tests: number;
  failed_tests: number;
  critical_findings: number;
  average_risk_score: number;
  progress_percent: number;

  started_at?: string | null;
  completed_at?: string | null;
};

export type InputEvaluation = {
  risk_level: string;
  detected_attack_types: string[];
  attack_intent: string;
  confidence: string;
  safe_to_run_in_sandbox: boolean;
  reason: string;
};

export type ModelResponse = {
  output?: string | null;
  latency_ms?: number | null;
  tokens_used?: number | null;
  estimated_cost?: number | null;
  error?: string | null;
  tool_calls?: unknown;
};

export type OutputEvaluation = {
  attack_success: boolean;
  refusal_quality?: string | null;
  sensitive_data_leaked: boolean;
  unsafe_output_detected: boolean;
  tool_misuse_detected: boolean;
  rag_injection_success: boolean;
  reasoning?: string | null;
  evidence?: string[];
  confidence?: string | null;
};

export type RiskAssessment = {
  risk_score: number;
  severity: string;
  confidence?: string | null;
  needs_human_review: boolean;
};

export type OwaspMappingItem = {
  id: string;
  name: string;
  reason: string;
};

export type SandboxReport = {
  test_id: string;
  platform: string;
  attack_scenario: string;
  provider: string;
  model: string;
  access_type: string;
  input_prompt: string;
  input_evaluation: InputEvaluation;
  model_response: ModelResponse;
  output_evaluation: OutputEvaluation;
  risk_assessment: RiskAssessment;
  owasp_mapping: OwaspMappingItem[];
  blue_team_recommendation: string[];
  final_status: string;
};

export type CampaignResult = {
  id: number;
  campaign_pk: number;
  campaign_id: string;

  scenario_pk?: number | null;
  scenario_id: string;
  attack_name: string;
  attack_category: string;
  severity: string;
  risk_level?: string;
  subcategory?: string | null;
  owasp_category?: string | null;

  model_name: string;
  mutation_type: string;

  input_prompt: string;
  mutated_prompt: string;

  sandbox_report?: SandboxReport | null;
  model_response?: ModelResponse | null;
  ai_evaluation?: OutputEvaluation | null;
  risk_assessment?: RiskAssessment | null;

  risk_score: number;
  final_status?: string | null;
  error_message?: string | null;

  created_at: string;
};

export type CampaignResultsListResponse = {
  campaign_id: string;
  total: number;
  items: CampaignResult[];
};
